Allow platform dependencies

kbx81 opened the pull request:
# What does this implement/fix?

Small PR that enables the ability to specify platform dependencies. For example:

```python
DEPENDENCIES = ["ota.esphome"]
```

...would require the platform `esphome` be present within the `ota` component configuration:

```yaml
ota:
  - platform: esphome
```

This enables more user-friendly validation errors should a required platform not be present.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

kbx81 committed:
Allow platform dependencies

nielsnl68 commented:
Nice idea @kbx81, is there a reason why you choose "esphome.ota" and not "ota.esphome"? To me the last version makes more sense to me. But maybe you have something else in mind that makes it easier to extend this later on?

kbx81 commented:
> Nice idea @kbx81, is there a reason why you choose "esphome.ota" and not "ota.esphome"? To me the last version makes more sense to me. But maybe you have something else in mind that makes it easier to extend this later on?

It was sort of arbitrary -- but it seems we use this arrangement for the logger tags, so I just copied that. It's trivial to change, so if we agree we should swap platform.component for component.platform, I've got no issues with that. 😄

kbx81 committed:
Switch to component.platform syntax, harden against config errors

kbx81 commented:
I have switched it to use `component.platform` syntax as I generally think it makes more sense here. Also added some additional logic to better handle various types of config errors.

kbx81 committed:
Clean-up

kbx81 committed:
Update `list-components.py`, too

kbx81 committed:
Make missing platform error message less confusing
